macro/src/generics.rs

```diff
@@ -1,65 +1,85 @@
-use crate::syntax::instantiate::NamedImplKey;
-use crate::syntax::resolve::Resolution;
 use crate::syntax::types::ConditionalImpl;
-use crate::syntax::{Impl, Lifetimes};
+use crate::syntax::{Lifetimes, Type, Types};
 use proc_macro2::TokenStream;
 use quote::ToTokens;
-use syn::{Lifetime, Token};
+use syn::Lifetime;
 
-pub(crate) struct ImplGenerics<'a> {
-    explicit_impl: Option<&'a Impl>,
-    resolve: Resolution<'a>,
-}
-
-pub(crate) struct TyGenerics<'a> {
-    key: &'a NamedImplKey<'a>,
-    explicit_impl: Option<&'a Impl>,
-    resolve: Resolution<'a>,
-}
-
-pub(crate) fn split_for_impl<'a>(
-    key: &'a NamedImplKey<'a>,
+/// Gets `(impl_generics, ty_generics)` pair that can be used when generating an `impl` for a
+/// generic type using something like:
+/// `quote! { impl #impl_generics SomeTrait for #inner #ty_generics }`.
+///
+/// Parameters:
+///
+/// * `inner` is the generic type argument (e.g. `T` in something like `UniquePtr<T>`)
+/// * `explicit_impl` corresponds to https://cxx.rs/extern-c++.html#explicit-shim-trait-impls
+pub(crate) fn get_impl_and_ty_generics<'a>(
+    inner: &'a Type,
     conditional_impl: &ConditionalImpl<'a>,
-    resolve: Resolution<'a>,
-) -> (ImplGenerics<'a>, TyGenerics<'a>) {
-    let impl_generics = ImplGenerics {
-        explicit_impl: conditional_impl.explicit_impl,
-        resolve,
-    };
-    let ty_generics = TyGenerics {
-        key,
-        explicit_impl: conditional_impl.explicit_impl,
-        resolve,
-    };
-    (impl_generics, ty_generics)
+    types: &'a Types,
+) -> (&'a Lifetimes, Option<&'a Lifetimes>) {
+    match conditional_impl.explicit_impl {
+        Some(explicit_impl) => {
+            let impl_generics = &explicit_impl.impl_generics;
+            (impl_generics, None /* already covered via `#inner` */)
+        }
+        None => {
+            // Check whether `explicit_generics` are present.  In the example below,
+            // there are not `explicit_generics` in the return type.
+            //
+            //     mod ffi {
+            //         unsafe extern "C++" {
+            //             type Borrowed<'a>;
+            //             fn borrowed(arg: &i32) -> UniquePtr<Borrowed>;
+            //         }
+            //     }
+            //
+            // But this could have also been spelled with `explicit_generics`:
+            //
+            //             fn borrowed<'a>(arg: &'a i32) -> UniquePtr<Borrowed<'a>>;
+            let explicit_generics = get_generic_lifetimes(inner);
+            if explicit_generics.lifetimes.is_empty() {
```

Comment:
This is incorrect in the case of `<>`. For example this compiles successfully before this PR:

```rust
#[cxx::bridge]
mod ffi {
    struct Borrowed<'a> {
        s: &'a str,
    }
    extern "Rust" {
        fn f(_: UniquePtr<Borrowed<>>);
    }
}

fn f(_: cxx::UniquePtr<ffi::Borrowed<>>) {}
```

but generates syntactically invalid code after:

```console
error: expected one of `!`, `+`, `::`, `where`, or `{`, found `<`
 --> src/main.rs:3:20
  |
3 |     struct Borrowed<'a> {
  |                    ^ unexpected token
...
7 |         fn f(_: UniquePtr<Borrowed<>>);
  |                                     - expected one of `!`, `+`, `::`, `where`, or `{`
```